tamr_unify_client/dataset/collection.py

```diff
@@ -171,6 +171,189 @@ def create_from_dataframe(
 
         return dataset
 
+    def create_from_geodataframe(
+        self, geodf, primary_key_name, dataset_name, geo_attr="geometry"
+    ):
+        """Creates a dataset in this collection with the given name, creates an attribute for each column in the `geodf`
+        (with `primary_key_name` as the key attribute), and upserts a record for each row of `geodf`.
+
+        Each attribute has the default type `ARRAY[STRING]`, besides the key attribute, which will have type `STRING`,
+        and the geometry attribute `geo_attr` which will have the type `GEOSPATIAL`.
+
+        This function attempts to ensure atomicity, but it is not guaranteed. If an error occurs while creating
+        attributes or records, an attempt will be made to delete the dataset that was created. However, if this
+        request errors, it will not try again.
+
+        :param geodf: The data to create the dataset with.
+        :type df: :class:`geopandas.geodataframe.GeoDataFrame`
+        :param primary_key_name: The name of the primary key of the dataset. Must be a column of `geodf`.
+        :type primary_key_name: str
+        :param dataset_name: What to name the dataset in Tamr. There cannot already be a dataset with this name.
+        :type dataset_name: str
+        :param geo_attr: The name of the geometry column within the `geodf`. Must be a column of `geodf`. Defaults
+            to `geometry`.
+        :type geo_attr: str
+        :returns: The newly created dataset.
+        :rtype: :class:`~tamr_unify_client.dataset.resource.Dataset`
+        :raises KeyError: If `primary_key_name` is not a column in `df`.
+        :raises CreationError: If a step in creating the dataset fails.
+        """
+        if primary_key_name not in geodf.columns:
+            raise KeyError(f"{primary_key_name} is not an attribute of the data")
+
+        creation_spec = {"name": dataset_name, "keyAttributeNames": [primary_key_name]}
+        try:
+            dataset = self.create(creation_spec)
+        except HTTPError:
+            raise CreationError("Dataset was not created")
+        # after this point, if a request fails, try to undo the change by deleting this dataset
+
+        attributes = dataset.attributes
+        for col in geodf.columns:
+            if col == primary_key_name:
+                # this attribute already exists, so don't create it again
+                continue
+
+            if col == geo_attr:
+                # this column has a different creation spec
+                attr_spec = {
+                    "name": geo_attr,
+                    "description": "",
+                    "type": {
```

Comment:
Let's take the `type` out and put it into a proper JSON file and just read it in here, rather than inlining the full type since it's so large.

Reply:
Yep, for now I went with consistency of `create_from_dataframe`, but obviously just reading the type from somewhere else would be much nicer.

Reply:
Included:

`from tamr_unify_client.attribute.specs.tamr_geometry_type import TAMR_GEOMETRY_TYPE`

Reply:
Instead of `attribute/spec/tamr_geometry_type.py`, could you encode this in an actual JSON file (not a Python file with a `dict` that corresponds to that JSON) and read in via `json.load(f)`?

We can structure as `attribute/types/tamr_geometry.json`.

Reply:
Also, only the actual `type` should be encoded in the JSON file, so the resulting code would look like:

```python
import json
# pretend we have `tuc` defined as a path pointing to `tamr_unify_client/` dir
with open(tuc / "attribute" / "types" / "tamr_geometry.json") as f:
    type = json.load(f)

for attr_name in geo_attr_names:
    attr = {
        "name": attr_name,
        "description": "",
        "type": type,
        ...
    }
```

Reply:
Sure, did it to allow easy importing and relative paths. 
Does poetry include a `pkg_resources` equivalent?

Reply:
Yes, see https://python-poetry.org/docs/pyproject/#include-and-exclude

Reply:
And we're back to why I detest the import functionality of Python. For now *real* work calls so I'll look at.